### PR TITLE
Implemented a user-triggered rebase feature

### PR DESCRIPTION
**Pull request by javierron**
Here is an implementation of the feature described on https://github.com/dabutvin/Imgbot/issues/352

This was implemented such that when a rebase is triggered, only the conflicted images and possibly new images would be re-compressed.

In order for this to work, the app should be updated to trigger the [issue_comment](https://developer.github.com/v3/activity/events/types/#issuecommentevent) webhook.

Authors: @s1grun and @javierron

### CODE REVIEW

**Comment by dabutvin** [comment]
awesome! I'll take a look and do some testing soon

**Comment by s1grun** [comment]
Hi @dabutvin, I hope you're well! 

We just wanted to check in regarding the status of the PR. Implementing this is a part of an assignment, in a course at KTH for our master program, that has the deadline on the 30th of April. No pressure though, but it would be so great to get feedback before that if possible. Thank you in advance!

**Comment by dabutvin** [comment]
@s1grun understood. I will do my best 😄

**Comment by s1grun** [comment]
Hi @dabutvin,

Just touching base, have you had any time to take a look at the PR?

File: CompressImagesFunction/CompressImages.cs [hunk 1]
```diff
@@ -134,7 +133,53 @@ public static bool Run(CompressimagesParameters parameters, ICollector<CompressI
...
+            List<string> addedOrModifiedImagePaths = new List<string>();
+            List<string> deletedImagePaths = new List<string>();
+            if (parameters.IsRebase)
+            {
+                var refspec = string.Format("{0}:{0}", KnownGitHubs.BranchName);
+                Commands.Fetch(repo, "origin", new List<string> { refspec }, null, "fetch");
+
+                var diff = repo.Diff.Compare<TreeChanges>(repo.Branches[KnownGitHubs.BranchName].Commits.ElementAt(1).Tree, repo.Head.Tip.Tree);
+
+                if (diff == null)
```
Comment (dabutvin) [inline CompressImagesFunction/CompressImages.cs:146]: i think we should do something to validate that the head commit in this branch is actually made by imgbot.
I feel that it is important that we never modify anyone else's commits and we only amend commits made by the bot.

This is the scenario I am thinking about:
1. Imgbot opens a branch and adds a commit
2. User adds a new commit on top of imgbot's commit
3. This rebase runs and amends the user's commit

**Comment by javierron** [comment]
Hey! Thanks for the review! 

We will address your feedback soon and update the PR accordingly.

👍🏻

File: CompressImagesFunction/CompressImages.cs [hunk 2]
```diff
@@ -216,7 +336,12 @@ public static bool Run(CompressimagesParameters parameters, ICollector<CompressI
             }
             else
             {
-                repo.Network.Push(remote, $"refs/heads/{KnownGitHubs.BranchName}", new PushOptions
+                var refs = $"refs/heads/{KnownGitHubs.BranchName}";
+                if (parameters.IsRebase)
+                    refs = refs.Insert(0, "+");
```
[the remark this answers is not in the document]
Comment (javierron) [reply CompressImagesFunction/CompressImages.cs:341]: Yes!

>  To force a push to only one branch, use a + in front of the refspec to push

https://git-scm.com/docs/git-push

File: CompressImagesFunction/CompressImagesFunction.cs [hunk 3]
```diff
@@ -116,6 +115,7 @@ public static class CompressImagesFunction
                 Password = installationToken.Token,
                 RepoName = compressImagesMessage.RepoName,
                 RepoOwner = compressImagesMessage.Owner,
+                IsRebase = compressImagesMessage.IsRebase,
```
[the remark this answers is not in the document]
Comment (javierron) [reply CompressImagesFunction/CompressImagesFunction.cs:118]: This makes sense from a more general point of view than that in the issue describing this feature.

File: CompressImagesFunction/CompressImages.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dabutvin) [reply CompressImagesFunction/CompressImages.cs:341]: very cool 👍

File: CompressImagesFunction/CompressImages.cs [hunk 4]
```diff
@@ -155,6 +200,81 @@ public static bool Run(CompressimagesParameters parameters, ICollector<CompressI
...
+                    foreach (var item in status)
+                    {
+                        if (item.State == FileStatus.Conflicted)
+                        {
+                            Commands.Stage(repo, item.FilePath);
+                        }
+                    }
+
+                    repo.Commit(commitMessage, signature, signature);
+                }
```
[the remark this answers is not in the document]
Comment (javierron) [reply CompressImagesFunction/CompressImages.cs:235]: OK, removing this for now 👍

For optimizing the whole thing, what do you think about a checksum-based solution? We could store all (filename,checksum) entries on a table, and compare every time the compression function is called, therefore compressing only the images in which the checksum does not match.

In terms of operation cost, this solution would mean a storage/processing trade-off. I believe in the long run, and specially in repositories with hundreds/thousands of images the trade-off would be justified.

[the remark this answers is not in the document]
Comment (dabutvin) [reply CompressImagesFunction/CompressImages.cs:235]: that's a cool idea 👍 and i think you're right, could be a huge win in the long run.
Not sure how familiar you with azure storage, but ideally we could pull this off with using either blobs or tables https://docs.microsoft.com/en-us/azure/storage/common/storage-introduction in a performant way.

**Comment by javierron** [comment]
@dabutvin The changes you requested are now uploaded 😎

**Comment by javierron** [comment]
Any update on this? I'll be glad to keep working on this and get it merged if it's still relevant.

**Comment by dabutvin** [comment]
thanks for the reminder @javierron 
I totally let this slip. I'll take a look soon 👍

**Pull request merged**